File: frontend/apps/h5/src/lib/public-url.ts
const MANAGED_PUBLIC_PATH =
  /^\/api\/v1\/files\/public\/[A-Za-z0-9][A-Za-z0-9._~!$&()*+,;=:@-]*(?:\/[A-Za-z0-9][A-Za-z0-9._~!$&()*+,;=:@-]*)*$/;

function isPrivateIpv4(host: string): boolean {
  const octets = host.split(".").map(Number);
  if (
    octets.length !== 4 ||
    octets.some((value) => !Number.isInteger(value) || value < 0 || value > 255)
  ) {
    return false;
  }
  const [a, b] = octets;
  return (
    a === 0 ||
    a === 10 ||
    a === 127 ||
    (a === 169 && b === 254) ||
    (a === 172 && b >= 16 && b <= 31) ||
    (a === 192 && b === 168) ||
    a >= 224
  );
}

/** Defense-in-depth for persisted URLs rendered on the public consumer page. */
export function safePublicUrl(value: string | null | undefined): string | null {
  const normalized = value?.trim();
  if (!normalized) return null;
  if (MANAGED_PUBLIC_PATH.test(normalized)) return normalized;

  try {
    const parsed = new URL(normalized);
    const host = parsed.hostname.replace(/\.$/, "").toLowerCase();
    if (parsed.protocol !== "https:" || parsed.username || parsed.password)
      return null;
    if (
      !host.includes(".") ||
      host === "localhost" ||
      host.endsWith(".localhost") ||
      host.endsWith(".local")
    ) {
      return null;
    }
    if (
      isPrivateIpv4(host) ||
      host === "::1" ||
      host.startsWith("fe80:") ||
      host.startsWith("fc") ||
      host.startsWith("fd")
    ) {
      return null;
    }
    return normalized;
  } catch {
    return null;
  }
}

File: backend/app/utils/public_url.py
import ipaddress
import re
from urllib.parse import urlsplit

_MANAGED_PUBLIC_PATH = re.compile(
    r"^/api/v1/files/public/[A-Za-z0-9][A-Za-z0-9._~!$&()*+,;=:@-]*"
    r"(?:/[A-Za-z0-9][A-Za-z0-9._~!$&()*+,;=:@-]*)*$"
)


def normalize_public_url(value: str) -> str:
    """Return a canonical consumer-safe public URL or raise ``ValueError``.

    Trust-master data is rendered directly by the consumer H5. Keep the
    persisted contract intentionally narrow: managed public-file paths or
    public HTTPS URLs without embedded credentials.
    """

    if not isinstance(value, str):
        raise ValueError("Public URL must be a string")

    normalized = value.strip()
    if _MANAGED_PUBLIC_PATH.fullmatch(normalized):
        return normalized

    parsed = urlsplit(normalized)
    if parsed.scheme.lower() != "https" or not parsed.hostname:
        raise ValueError("Public URL must use HTTPS or a managed public-file path")
    if parsed.username is not None or parsed.password is not None:
        raise ValueError("Public URL must not contain credentials")

    host = parsed.hostname.rstrip(".").lower()
    if host == "localhost" or host.endswith(".localhost") or host.endswith(".local") or "." not in host:
        raise ValueError("Public URL host must be publicly routable")
    try:
        address = ipaddress.ip_address(host)
    except ValueError:
        pass
    else:
        if not address.is_global:
            raise ValueError("Public URL host must be publicly routable")
    return normalized
